azurelinuxagent/common/protocol/wire.py

```diff
@@ -995,6 +1075,41 @@ def check_wire_protocol_version(self):
                      "advised by Fabric.").format(PROTOCOL_VERSION)
             raise ProtocolNotFoundError(error)
 
+    def send_request_hostplugin_first(self, direct_func, host_func):
+        # For certain calls we want them to go through the HostGAPlugin first, since it
+        # supports the If-None-Match header that reduces how often we need to parse
+
+        # note that this method requires a lambda that returns two values
+        ret = None
+        etag = None
+        not_modified = False
+        try:
+            ret, etag, not_modified = host_func()
+
+            # Different direct channel functions report failure in different ways: by returning None, False,
+            # or raising ResourceGone or InvalidContainer exceptions.
+            if not ret and not not_modified:
+                logger.info("Request failed using the HostGAPlugin. switching to direct.")
+        except (ResourceGoneError, InvalidContainerError) as e:
+            msg = "Request failed with the current host plugin configuration." \
+                  "ContainerId: {0}, role config file: {1}." \
+                  "Error: {2}".format(self.host_plugin.container_id, self.host_plugin.role_config_name, ustr(e))
+            logger.info(msg)
+
+        if ret or not_modified:
+            return ret, etag
+
+        try:
+            ret = direct_func()
+        except (ResourceGoneError, InvalidContainerError) as e:
+            logger.info("Request failed with direct. Error: {0}".format(ustr(e)))
+
+        if not ret:
+            # Try through the host one more time
```

Comment:
Given that we call this every 3 secs and that our restutil lib has its own retry logic, the 3rd retry might cause a lot of time delay in case of some network failure (not to mention the verbosity in our logs). I wonder if it would cause more harm than good?